[cssom-view-1] Added definition of getBoxQuads()

jogibear9988 opened the pull request:
Fixes #10537

A linked issue:
#10514

A firefox issue with a little bit of description:
https://bugzilla.mozilla.org/show_bug.cgi?id=1107559

A polyfill for the API
https://github.com/jogibear9988/getBoxQuadsPolyfill

jogibear9988 committed:
Work on description of getBoxQuads API

jogibear9988 committed:
Update Overview.bs

jogibear9988 committed:
Update Overview.bs

jogibear9988 commented:
One issues atm:
My polyfill only works for 2D transformations, 3D with perspective do not work (yet). Have to figure out how to do this. Cause of that, also my spec only works for 2D. I've found some spect for the 3D matrix, wich then also could maybe reused: https://drafts.csswg.org/css-transforms-2/#accumulated-3d-transformation-matrix-computation

SebastianZ commented:
I think it makes more sense to add another option to `getBoxQuads()` to consider the transformed coordinates instead of introducing a separate function `getCompleteTransform()` for that.

I imagine something like `getBoxQuads({considerTransformations: true})`.

Sebastian

jogibear9988 commented:
> One issues atm: My polyfill only works for 2D transformations, 3D with perspective do not work (yet). Have to figure out how to do this. Cause of that, also my spec only works for 2D. I've found some spect for the 3D matrix, wich then also could maybe reused: https://drafts.csswg.org/css-transforms-2/#accumulated-3d-transformation-matrix-computation

should be fixed in current polyfill

jogibear9988 commented:
> I think it makes more sense to add another option to `getBoxQuads()` to consider the transformed coordinates instead of introducing a separate function `getCompleteTransform()` for that.
> 
> I imagine something like `getBoxQuads({considerTransformations: true})`.
> 
> Sebastian

it's in a extra method, cause it is used from the 3 other functions in the spec

emilio commented on cssom-view-1/Overview.bs:
```diff
@@ -1781,46 +1781,76 @@ Document includes GeometryUtils;
 
-The <dfn method for=GeometryUtils lt="getBoxQuads(options)|getBoxQuads()">getBoxQuads(<var>options</var>)</dfn> method must run the following steps:
-
+The helper method getCompleteTransform(anchestor) must run the following steps:
```
You need to get the element as a parameter, and use `<div algorithm>` for this for example:

```
<div algorithm="get the complete transform">
  In order to <dfn export>get the complete transform</dfn> of an node <var>node</var> relative to a node <var>ancestor</ancestor> ...
</div>

jogibear9988 replied:
I don't get what you mean with this....
getCompleteTransform has a parameter, wich is anchestor

emilio commented on cssom-view-1/Overview.bs:
```diff
@@ -1781,46 +1781,76 @@ Document includes GeometryUtils;
-
-  test block in inline
+1. let currentElement be the element getCompleteTransform() called on.
+2. let transformationMatrix be the current transformation matrix of currentElement (https://www.w3.org/TR/css-transforms-2/#ctm)
```
Please use bikeshed references for this.

emilio commented on cssom-view-1/Overview.bs:
```diff
@@ -1781,46 +1781,76 @@ Document includes GeometryUtils;
 
 <ol>
- <li><p class=issue>...
+1. get the resulting matrix between the element and the document.defaultView via the algorithm defined in getCompleteTransform
```
All these algorithms need to be in terms of `node`s, as `GeometryUtils` also applies to text nodes and such.

jogibear9988 replied:
Is node correct? Or should it be "GeometryNode"?
<img width="757" height="87" alt="image" src="https://github.com/user-attachments/assets/b7dd425e-8d78-4dce-955d-bd01b5c1b483" />

emilio commented on cssom-view-1/Overview.bs:
```diff
@@ -1781,46 +1781,76 @@ Document includes GeometryUtils;
+   - 'border': no change
+   - 'margin': x - margin.left, y - margin.top
+   - 'padding': x + border.leftWidth, y + border.topWidth
+   - 'content': x + border.leftWidth + padding.left, y + border.topWidth + padding.top
```
This is not actually based on the computed style but on the actual used padding box etc. Those boxes are defined already elsewhere and we should reuse the definitions.

Psychpsyo replied:
> Yeah, but I need to link to the Algorythm how they are calculated, or is this enough?

There is no algorithm like that for these. They are defined in English instead. Browsers have algorithms internally that match those descriptions and can call into those when a spec says to do something with one of these boxes, so linking to their definitions is enough to unambiguously tell an implementer what needs to happen.

More modern specs like to define their concepts in terms of algorithms to derive them, because that makes implementation easier and leaves less room for creative (mis-)interpretation of the definition.

TL;DR: You want to link to a concept's definition when using it. Whether that is an algorithm or not doesn't matter.

emilio reviewed (changes requested):
Thanks, this looks like a good start. Please look at how other specs (specially recent ones) use bikeshed in order to get the formatting and references correct. Also this needs to work in terms of nodes and boxes, not elements.

jogibear9988 committed:
fix duplicated ol

jogibear9988 committed:
Merge branch 'main' of https://github.com/w3c/csswg-drafts into patch-1

jogibear9988 committed:
fixes

jogibear9988 committed:
fixes

jogibear9988 committed:
adjust spec

jogibear9988 commented:
@emilio 
is it okay like now?

emilio reviewed (changes requested):
This still needs a bit of work to get the spec into a reasonable shape, but looks in the right direction. Thanks!

whimboo commented:
Hi @jogibear9988. Would you maybe have the time to update this PR based on the review feedback from Emilio? Thanks a lot!

jogibear9988 commented:
Sorry, I've read them, and hopefully fix in the next days

whimboo commented:
Hi @jogibear9988. I wanted to check back with you about the status of this PR. Is it still something you can continue working on? Thanks!

jogibear9988 commented:
@whimboo 
sorry, had not much time to work on this.
at the moment I fix my polyfill (https://github.com/jogibear9988/getBoxQuadsPolyfill) to also respect offset-.... css properties. Didn't know they exist. 
Also there is a small issue with webcomponents I try to fix.
Then I look to fix this spec.

bramus commented:
> Also I think the spec needs an upgrade to also include a function to calculate the complete transformation matrix between two elements. My polyfill does this, and I need this also some time in my designer, and this is not possible only with the boxQuads result. is this something I also should add here, or should this be an extra issue/pull req?

This would need a separate PR – one that would need a CSS WG Resolution first to get the group’s consensus to adding it.

whimboo commented:
Hello @jogibear9988. I want to check back with you if you may have some time that this PR could be moved forward? Or is it blocked based on the last comment? Thanks!

jogibear9988 committed:
Merge branch 'w3c:main' into patch-1

Copilot committed:
Initial plan

Address Emilio's review feedback on GeometryUtils spec definitions

Rewrite getCompleteTransform, getBoxQuads, convertQuadFromNode,
convertRectFromNode, and convertPointFromNode algorithm definitions
using proper bikeshed markup:

- Use <div algorithm> block with <dfn> for getCompleteTransform
- Use <var> tags for all variables throughout
- Use {{}} bikeshed notation for DOM types (DOMQuad, DOMMatrix, etc.)
- Reference css-transforms-1 CTM definition properly
- Handle Document (layout viewport) vs Element/Text (box fragments)
- Use proper <a spec=css-break>box fragment</a> references
- Reference margin/padding/border/content edges correctly
- Fix relativeTo (not realtiveTo) and use layout viewport as fallback
- Give meaningful variable names (thisTransformToViewport, etc.)
- Link convertQuadFromNode to convertPointFromNode explicitly
- Reference used values of CSS properties for box adjustments

Co-authored-by: jogibear9988 <364896+jogibear9988@users.noreply.github.com>

Address code review: add detailed box adjustment steps and non-invertible matrix note

- Spell out how margin/padding/content box adjustments work using
  specific CSS property used values
- Add note about non-invertible transformation matrices
- Add link-defaults for border-right-width and border-bottom-width

Co-authored-by: jogibear9988 <364896+jogibear9988@users.noreply.github.com>

Initial plan

Fix bikeshed reference: use [=list=] in getBoxQuads algorithm

Co-authored-by: jogibear9988 <364896+jogibear9988@users.noreply.github.com>
Agent-Logs-Url: https://github.com/jogibear9988/csswg-drafts/sessions/2cc807ab-9305-4dd8-b0f7-dafb7f3c6be2

Work on description of getBoxQuads API

Update Overview.bs

Update Overview.bs

fix duplicated ol

fixes

fixes

adjust spec

Initial plan

I updated the GeometryUtils section to address the review feedback.

The main changes are:
- rewrote getBoxQuads() in terms of nodes and box fragments, with layout-viewport coordinates as the default when relativeTo is omitted
- added explicit Document and Text handling
- moved the box-space wording to helper algorithms and removed the margin/padding/content edge arithmetic based on used values
- reused convertQuadFromNode()/convertPointFromNode() for the actual coordinate conversion path

I also validated the edited file locally. A remote Bikeshed run still hits unrelated existing markdown errors later in cssom-view-1, but it did not report new fatal issues in the GeometryUtils block.

fixes after review

Copilot committed:
Initial plan

Copilot committed:
Address Emilio's review feedback on GeometryUtils spec definitions

Rewrite getCompleteTransform, getBoxQuads, convertQuadFromNode,
convertRectFromNode, and convertPointFromNode algorithm definitions
using proper bikeshed markup:

- Use <div algorithm> block with <dfn> for getCompleteTransform
- Use <var> tags for all variables throughout
- Use {{}} bikeshed notation for DOM types (DOMQuad, DOMMatrix, etc.)
- Reference css-transforms-1 CTM definition properly
- Handle Document (layout viewport) vs Element/Text (box fragments)
- Use proper <a spec=css-break>box fragment</a> references
- Reference margin/padding/border/content edges correctly
- Fix relativeTo (not realtiveTo) and use layout viewport as fallback
- Give meaningful variable names (thisTransformToViewport, etc.)
- Link convertQuadFromNode to convertPointFromNode explicitly
- Reference used values of CSS properties for box adjustments

Co-authored-by: jogibear9988 <364896+jogibear9988@users.noreply.github.com>

Copilot committed:
Address code review: add detailed box adjustment steps and non-invertible matrix note

- Spell out how margin/padding/content box adjustments work using
  specific CSS property used values
- Add note about non-invertible transformation matrices
- Add link-defaults for border-right-width and border-bottom-width

Co-authored-by: jogibear9988 <364896+jogibear9988@users.noreply.github.com>

Copilot committed:
Initial plan

Copilot committed:
Fix bikeshed reference: use [=list=] in getBoxQuads algorithm

Co-authored-by: jogibear9988 <364896+jogibear9988@users.noreply.github.com>
Agent-Logs-Url: https://github.com/jogibear9988/csswg-drafts/sessions/2cc807ab-9305-4dd8-b0f7-dafb7f3c6be2

jogibear9988 committed:
Merge pull request #3 from jogibear9988/copilot/sub-pr-2

[cssom-view] Specify GeometryUtils algorithms (getBoxQuads, convertQuadFromNode, convertRectFromNode, convertPointFromNode)

Copilot committed:
Initial plan

Copilot committed:
Merge remote-tracking branch 'origin/patch-1' into copilot/fix-issues-from-pull-10538

# Conflicts:
#	cssom-view-1/Overview.bs

jogibear9988 committed:
Merge pull request #5 from jogibear9988/copilot/sub-pr-2

Resolve merge conflicts with updated patch-1 base branch

jogibear9988 committed:
Merge pull request #2 from jogibear9988/copilot/fix-issues-from-pull-10538

Copilot/fix issues from pull 10538

jogibear9988 committed:
I updated the GeometryUtils section to address the review feedback.

The main changes are:
- rewrote getBoxQuads() in terms of nodes and box fragments, with layout-viewport coordinates as the default when relativeTo is omitted
- added explicit Document and Text handling
- moved the box-space wording to helper algorithms and removed the margin/padding/content edge arithmetic based on used values
- reused convertQuadFromNode()/convertPointFromNode() for the actual coordinate conversion path

I also validated the edited file locally. A remote Bikeshed run still hits unrelated existing markdown errors later in cssom-view-1, but it did not report new fatal issues in the GeometryUtils block.

jogibear9988 committed:
Merge branch 'w3c:main' into patch-1

jogibear9988 commented on cssom-view-1/Overview.bs:
```diff
@@ -2192,26 +2194,107 @@ Document includes GeometryUtils;
+        {{CSSBoxType/"margin"}},
+        adjust <var>rect</var> to describe the <a>margin edge</a> of the fragment
+        by subtracting the used value of 'margin-left' from {{DOMRect/x}},
+        subtracting the used value of 'margin-top' from {{DOMRect/y}},
```
Thanks, that makes sense.

I changed this in the latest update. `getBoxQuads()` now defines the fragment rect in local coordinates with the border edge at 0,0, and for the `"margin"`, `"padding"`, and `"content"` cases it refers to the fragment’s corresponding box area instead of re-deriving those edges from used values.

So this should now follow the standard box definitions rather than trying to reconstruct them from first principles.

jogibear9988 commented:
also added some description to my polyfill: https://github.com/jogibear9988/getBoxQuadsPolyfill/blob/main/csswg-getBoxQuads-calculation.md

jogibear9988 commented:
I also have a new library wich uses getBoxQuads API, to be able to create screenshots or vector exports of HTML pages: https://github.com/node-projects/layout2vector

emilio commented on cssom-view-1/Overview.bs:
```diff
@@ -2177,7 +2179,7 @@ The <dfn attribute for=MouseEvent>offsetY</dfn> attribute must follow these step
 dictionary BoxQuadOptions {
   CSSBoxType box = "border";
-  GeometryNode relativeTo; // XXX default document (i.e. viewport)
+    GeometryNode relativeTo; // if omitted, coordinates are relative to the layout viewport
```
nit: Indentation. Also the "if omitted.." should be in the spec prose

emilio replied:
It's already, so just remove the comment.

emilio commented on cssom-view-1/Overview.bs:
```diff
@@ -2189,7 +2191,7 @@ interface mixin GeometryUtils {
   DOMQuad convertQuadFromNode(DOMQuadInit quad, GeometryNode from, optional ConvertCoordinateOptions options = {});
   DOMQuad convertRectFromNode(DOMRectReadOnly rect, GeometryNode from, optional ConvertCoordinateOptions options = {});
-  DOMPoint convertPointFromNode(DOMPointInit point, GeometryNode from, optional ConvertCoordinateOptions options = {}); // XXX z,w turns into 0
+    DOMPoint convertPointFromNode(DOMPointInit point, GeometryNode from, optional ConvertCoordinateOptions options = {}); // result is flattened to 2D
```
nit: Ditto

emilio commented on cssom-view-1/Overview.bs:
```diff
@@ -2200,26 +2202,158 @@ Document includes GeometryUtils;
-  cross-frames not allowed, throw WrongDocumentError?
+The {{GeometryUtils}} methods operate on a node's local box geometry.
+The {{GeometryUtils/getBoxQuads()}} method returns one {{DOMQuad}} per relevant
+box fragment, in content order.
```
This should be in flattened tree order, and should link to the definition.

emilio commented on cssom-view-1/Overview.bs:
```diff
@@ -2200,26 +2202,158 @@ Document includes GeometryUtils;
+    1. Let <var>transformationMatrix</var> be a new {{DOMMatrix}}.
+    1. While <var>current</var> is not null:
+        1. If <var>current</var> is <var>ancestor</var>, return <var>transformationMatrix</var>.
+        1. Let <var>parent</var> be null.
```
Parent is not `parent` right? It's more of `ancestor`

emilio reviewed (changes requested):
Thanks, this looks a lot better! Some nits and tweaks but looks good to me with those fixed.

emilio reviewed (approved):
I think this is in a pretty good state now. Could you squash please? Also I don't know what needs to happen to make the IPR bot happy... @svgeesus can you help?

emilio reviewed (approved):
Let's try this, thanks!
